src/dir.rs

```diff
@@ -0,0 +1,167 @@
+use {Error, Result};
+use errno::Errno;
+use libc;
+use std::os::unix::io::{AsRawFd, IntoRawFd, RawFd};
+use std::{ffi, fmt, ptr};
+
+#[cfg(target_os = "linux")]
+use libc::{dirent64 as dirent, readdir64_r as readdir_r};
+
+#[cfg(not(target_os = "linux"))]
+use libc::{dirent, readdir_r};
+
+/// An open directory.
+///! This is a lower-level interface than `std::fs::ReadDir`. Notable differences:
+///!    * can be opened from a file descriptor (as returned by `openat`, perhaps before knowing
+///!      if the path represents a file or directory).
+///!    * implements `AsRawFd`, so it can be passed to `fstat`, `openat`, etc.
+///!    * can be iterated through multiple times without closing and reopening the file
```

Comment:
How can it be iterated through multiple times if it only implemented `IntoIterator` and not `Iterator`?

Reply:
`&mut Directory` implements `IntoIterator`, so you can get an iterator more than once. The test does this.

I could have `Dir` implement `Iterator` instead but you'd have to remember to rewind it.

Reply:
Oh, I see.  You implemented `IntoIterator` for `&Dir`, not `Dir`.  I think it would be more idiomatic instead for `Dir` to have a `iter()` method, that returns an object which implemented `Iterator`.